File: Python/DoublyLinkedList.py
class Node:
    def __init__(self, item):
        self.item = item
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.first = Node(None)
        self.last = Node(None)
        self.num = 0
        self.first.next = self.last
        self.last.prev = self.first

    # true if list is empty else false
    def isEmpty(self):
        """
        return True if the LL is empty else returns False
        """
        return self.num == 0

    # returns number of elements
    def size(self):
        """
        return the integer size of the list
        """
        return self.num

    # prepends element to the collection
    def addFirst(self, item):
        """
        Input: item
        Action: adds item to first position (0 index) of the list
        """
        if (item == None): 
            raise Exception("Illegal argument")

        newNode = Node(item)
        newNode.next = self.first.next
        newNode.prev = self.first
        self.first.next.prev = newNode
        self.first.next = newNode
        self.num += 1

    # appends element to the collection
    def addLast(self, item):
        """
        Input: item
        Action: adds item to the last position of the list
        """
        if (item == None):
            raise Exception("Illegal argument")

        newNode = Node(item)
        newNode.next = self.last
        newNode.prev = self.last.prev
        self.last.prev.next = newNode
        self.last.prev = newNode
        self.num += 1

    # Inserts element at specified position
    def add(self, index, item):
        """
        Inputs: an integer index, item
        Action: adds the item to the specified index in the list if the
                index in not out of bounds
        """
        if (index >= self.num or index < 0):
            raise Exception("Index out of bounds")
        if (item == None):
            raise Exception("Illegal input element")

        idx = 0
        current = Node(item)
        current.next = self.first.next
        
        while (idx != index):
            current.next = current.next.next
            idx += 1
        
        current.prev = current.next.prev
        current.prev.next.prev = current
        current.prev.next = current
        self.num += 1

    # removes first element of the list
    def removeFirst(self):
        """
        removes and returns the first element of the list (index = 0)
        """
        if (self.num == 0):
            raise Exception("List is empty")

        delNode = self.first.next
        out = delNode.item
        self.first.next = delNode.next
        self.first.next.prev = self.first
        self.num -= 1
        return out

    # removes last element of the list
    def removeLast(self):
        """
        removes and return the last element of the list
        """
        if (self.num == 0):
            raise Exception("List is empty")

        delNode = self.last.prev
        out = delNode.item
        self.last.prev = delNode.prev
        self.last.prev.next = self.last
        self.num -= 1
        return out

    # clears all elements of list
    def clear(self):
        """
        clears all the elements from the list
        """
        self.first = Node(None)
        self.last = Node(None)
        self.num = 0
        self.first.next = self.last
        self.last.prev = self.first

    # creates and return identical list
    def clone(self):
        """
        creates and returns a new identical linked list
        """
        newLL = DoublyLinkedList()
        count = 0
        current = Node(None)
        current.next = self.first.next

        while (count != self.num):
            item = current.next.item
            newLL.addLast(item)
            current.next = current.next.next
            count += 1

        return newLL

    # checks if the list has the specified element
    def contains(self, item):
        """
        Input: item
        Output: returns True if the Linked list has the input item
                else returns False
        """
        if (item == None):
            raise Exception("Illegal argument")

        current = self.first
        idx = 0

        while (idx != self.num):
            current = current.next
            element = current.item
            idx += 1
            if (element == item):
                return True

        return False
    
    # returns first element of the list without removing it
    def element(self):
        """
        return the first element of the list (index = 0)
        """
        if (self.num == 0):
            raise Exception("List is empty")
        return self.first.next.item

    # returns element at a specified position
    def get(self, index):
        """
        Input: an integer index
        Output: return the item fromt he list at the input index
                if the index is not out of bounds
        """
        if (index >= self.num):
            raise Exception("index out of bounds")

        idx = 0
        current = self.first.next

        while (idx != index):
            current = current.next
            idx += 1

        item = current.item
        return item

    # returns first element
    def getFirst(self):
        """
        returns the first element of list without mutation (index = 0)
        """
        if (self.num == 0):
            raise Exception("List is empty")
        return self.first.next.item

    # returns last element
    def getLast(self):
        """
        returns the last element of the list without mutation
        """
         if (self.num == 0):
             raise Exception("List is empty")
         return self.last.prev.item

    # returns the index of first occurence of specified element or -1 if no such element
    def indexOf(self, item):
        """
        Input: item
        Output: returns the index of the first occurence of the input item or
                -1 if no such item is present in the list
        """
        if (self.num == 0):
            return -1
        idx = 0
        current = self.first.next

        while (idx != self.num):
            currItem = current.item
            if (currItem == item):
                return idx
            current = current.next
            idx += 1

        return -1

    # iterates the list
    def iterate(self):
        """
        yields elements of the list according to index order
        """
        if (self.num == 0):
            raise Exception("List is empty")
        current = Node(None)
        current.next = self.first.next
        while (current.next != self.last):
            item = current.next.item
            current.next = current.next.next
            yield item

    # returns a python list
    def toList(self):
        """
        return a python list identical to the linked list
        """
        if (self.num == 0):
            raise Exception("List is empty")
        out = []
        for elem in self.iterate():
            out.append(elem)
        return out

File: Java/DoublyLinkedList.java
import java.util.NoSuchElementException;
import java.util.Iterator;

public class DoublyLinkedList<Item> implements Iterable<Item> {

    private Node first;
    private Node last;
    private int num;

    private class Node {

        Item item;
        Node next;
        Node prev;

    }
    // Initializes an empty Linked list
    public DoublyLinkedList() {

        this.first = new Node();
        this.last = new Node();
        this.num = 0;
        first.item = null;
        last.item = null;
        first.next = last;
        last.prev = first;

    }

    // true if list is empty else false
    public boolean isEmpty() {

        return (this.num == 0);

    }

    // returns number of elements
    public int size() {
 
        return num;

    }

    // prepends element to the collection
    public void addFirst(Item item) {

        if (item == null) throw new IllegalArgumentException("null argument found!");

        Node newNode = new Node();
        newNode.item = item;
        newNode.next = this.first.next;
        newNode.prev = this.first;
        this.first.next.prev = newNode;
        this.first.next = newNode;
        this.num++;

    }

    // appends element to the collection
    public void addLast(Item item) {

        if (item == null) throw new IllegalArgumentException("null argument found!");

        Node newNode = new Node();
        newNode.item = item;
        newNode.next = this.last;
        newNode.prev = this.last.prev;
        this.last.prev.next = newNode;
        this.last.prev = newNode;
        this.num++;

    }

    // Inserts Item at specified position
    public void add(int index, Item item) {

        if (index >= this.num || index < 0) {
            
            throw new IndexOutOfBoundsException("index is out of bounds!");

        }

        if (item == null) throw new IllegalArgumentException("null argument found!");

        int idx = 0;
        Node current = new Node();
        current.item = item;
        current.next = this.first.next;
        
        while (idx != index) {

            current.next = current.next.next;
            idx++;

        }

        current.prev = current.next.prev;
        current.prev.next.prev = current;
        current.prev.next = current;
        this.num++;

    }

    // appends element to the list
    public void add(Item item) {

        if (item == null) throw new IllegalArgumentException("null argument found!");
        addLast(item);

    }

    // removes first element of list
    public Item removeFirst() {

        if (this.num == 0) throw new NoSuchElementException("No element found");

        Node delNode = this.first.next;
        Item out = delNode.item;
        this.first.next = delNode.next;
        this.first.next.prev = this.first;
        this.num--;
        return out;

    }

    // removes last element of list
    public Item removeLast() {

        if (this.num == 0) throw new NoSuchElementException("No element found");
        Node delNode = this.last.prev;
        Item out = delNode.item;
        this.last.prev = delNode.prev;
        this.last.prev.next = this.last;
        this.num--;
        return out;

    }

    // clears all elements of list
    public void clear() {

        this.first = new Node();
        this.last = new Node();
        this.num = 0;
        first.item = null;
        last.item = null;
        first.next = last;
        last.prev = first;

    }

    // creates and retures a identical list
    public DoublyLinkedList<Item> clone() {

        DoublyLinkedList<Item> newLL = new DoublyLinkedList<Item>();
        int count = 0;
        Node current = new Node();
        current.next = this.first.next;
        
        while (count != this.num) {

            Item item = current.next.item;
            newLL.addLast(item);
            current.next = current.next.next;
            count++;
        }

        return newLL;

    }

    // checks if the list has the specified element
    public boolean contains(Item item) {

        if (item == null) throw new IllegalArgumentException("argument is null");

        Node current = this.first;
        int idx = 0;

        while (idx != this.num) {

            current = current.next;
            Item element = current.item;
            idx++;
            if (element.equals(item)) return true;

        }

        return false;

    }

    // returns the first element of the list without removing
    public Item element() {

        if (this.num == 0) throw new NoSuchElementException("List is Empty");
        return this.first.next.item;

    }

    // returns the element at the specified position
    public Item get(int index) {

        int idx = 0;
        Node current = this.first.next;

        while (idx != index) {

            current = current.next;
            idx++;

        }

        Item item = current.item;
        return item;

    }

    // returns first element
    public Item getFirst() {

        if (this.num == 0) throw new NoSuchElementException("List is empty");
        return this.first.next.item;

    }

    // returns last element
    public Item getLast() {

        if (this.num == 0) throw new NoSuchElementException("List is empty");
        return this.last.prev.item;

    }

    // returns the index of first occurence of specified element or -1 if no such element
    public int indexOf(Item item) {

        if (this.num == 0) return -1;
        int idx = 0;
        Node current = this.first.next;
        while (idx != this.num) {

            Item currItem = current.item;
            if (currItem.equals(item)) return idx;
            current = current.next;
            idx++;

        }

        return -1;

    }

    public Iterator<Item> iterator() {

        return new LinkedListIterator();

    }

    private class LinkedListIterator implements Iterator<Item> {

        private Node current = first;

        public boolean hasNext(){

            return current.next != last;

        }

        public Item next() {

            if (!hasNext()) throw new NoSuchElementException("no more elements");

            current = current.next;
            Item item = current.item;
            return item;

        }

        /*
        public void remove() {

        }
        */
    }

/*
    public static void main(String[] args) {

        // unit tests:
        DoublyLinkedList<String> words = new DoublyLinkedList<String>();
        System.out.println("Is the LL empty = "+words.isEmpty());
        System.out.println("LL size = "+words.size());
        words.addFirst("Ace");
        words.addLast("King");
        words.addLast("Queen");
        words.addFirst("Two");
        words.addFirst("One");
        words.addLast("Jack");
        System.out.println("LL size = "+words.size());
        System.out.println("Current List is : ");
        for (String s: words) { System.out.println(s); }
        words.removeFirst();
        words.removeLast();
        words.removeFirst();
        words.removeFirst();
        for (String s: words) { System.out.println(s); }
        System.out.println("LL size = "+words.size());
        words.add(1, "Ace");
        words.add("Five");
        for (String s: words) { System.out.println(s); }
        System.out.println("LL size = "+words.size());        
        
        DoublyLinkedList<String> newWords = words.clone();
        for (String s: newWords) { System.out.println(s); }
        newWords.clear();
        System.out.println("LL size = "+newWords.size());
        // for (String s: newWords) { System.out.println(s); }

        System.out.println(words.contains("Ace"));
        System.out.println(words.contains("Seven"));
        System.out.println(words.element());
        for (String s: words) { System.out.println(s); }
        System.out.println(words.get(3));
        System.out.println(words.get(2));
        System.out.println(words.getFirst());
        System.out.println(words.getLast());
        System.out.println(words.indexOf("Ace"));
        System.out.println(words.indexOf("Eight"));

    }
*/
}
